### PR TITLE
✨ DDD Add Better Domain Event Example

### PR DESCRIPTION
**Pull request by danielmackay**
﻿> 1. What triggered this change? (PBI link, Email Subject, conversation + reason, etc)

✏️ #264

> 2. What was changed?

✏️ Added an event handler to update Team Power, when an individual heroes power is updated

> 3. Did you do pair or mob programming?

✏️ No

### CODE REVIEW

File: src/Infrastructure/Persistence/Interceptors/DispatchDomainEventsInterceptor.cs [hunk 1]
```diff
@@ -7,18 +7,32 @@ namespace SSW.CleanArchitecture.Infrastructure.Persistence.Interceptors;
 
 public class DispatchDomainEventsInterceptor(IMediator mediator) : SaveChangesInterceptor
 {
-    public override InterceptionResult<int> SavingChanges(DbContextEventData eventData, InterceptionResult<int> result)
+    // NOTE: There are two options for dispatching domain events:
```
Comment (danielmackay) [inline src/Infrastructure/Persistence/Interceptors/DispatchDomainEventsInterceptor.cs:10]: @william-liebenberg @matt-goldman-ssw - you guys were looking into this some time back and may be interested in this decision.

File: src/Domain/Heroes/PowerLevelUpdatedEvent.cs [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
+using SSW.CleanArchitecture.Domain.Teams;
 
 namespace SSW.CleanArchitecture.Domain.Heroes;
 
 public record PowerLevelUpdatedEvent : DomainEvent
 {
     public HeroId Id { get; }
-    public string Name { get; }
-    public int PowerLevel { get; }
+    public TeamId? TeamId { get; }
```
Comment (drwharris) [inline src/Domain/Heroes/PowerLevelUpdatedEvent.cs:10]: Why not just store the Hero?

File: src/Domain/Teams/Team.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@
 namespace SSW.CleanArchitecture.Domain.Teams;
 
 // For strongly typed IDs, check out the rule: https://www.ssw.com.au/rules/do-you-use-strongly-typed-ids/
```
Comment (william-liebenberg) [inline src/Domain/Teams/Team.cs:8]: Did you update the rule for the 'sealed' use case?

File: src/Infrastructure/Persistence/Interceptors/DispatchDomainEventsInterceptor.cs (hunk 1 above)
Comment (william-liebenberg) [reply src/Infrastructure/Persistence/Interceptors/DispatchDomainEventsInterceptor.cs:10]: Still a pending effort on this one :)

**Review by william-liebenberg — APPROVED** [review]

File: src/Domain/Teams/Team.cs (hunk 3 above)
Comment (danielmackay) [reply src/Domain/Teams/Team.cs:8]: No I haven't updated the rule.  This change was a strange one.  The problem I've found is that sometimes you need to use optional/nullable foreign keys.  When you do this EF Core doesn't like trying to use nullable structs for FKs.  The only way I could find to fix this is to change from a `record struct` to `record`.

Should probably create a ticket to further investigte this and update the rule if necesary.

Comment (danielmackay) [reply src/Domain/Teams/Team.cs:8]: Added a PBI to track - https://github.com/SSWConsulting/SSW.CleanArchitecture/issues/329#issue-2326636333

File: src/Domain/Heroes/PowerLevelUpdatedEvent.cs (hunk 2 above)
Comment (danielmackay) [reply src/Domain/Heroes/PowerLevelUpdatedEvent.cs:10]: To capture the exact state of the hero when the changes were raised.  In our example it doesn't matter too much, but in other examples if many domain events were fired we would want them to be unique and NOT just point to the latest version of the Hero (which would all be the same)

**Pull request merged**